wgpu-core/src/lock/observing.rs

```diff
@@ -0,0 +1,330 @@
+//! Lock types that observe lock acquisition order.
+//!
+//! This module's [`Mutex`] type is instrumented to observe the
+//! nesting of `wgpu-core` lock acquisitions. Whenever `wgpu-core`
+//! acquires one lock while it is already holding another, we note
+//! that nesting pair. This tells us what the [`LockRank::followers`]
+//! set for each lock would need to include to accommodate
+//! `wgpu-core`'s observed behavior.
+//!
+//! When `wgpu-core`'s `observe_locks` feature is enabled, if the
+//! `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is set to the
+//! path of an existing directory, then every thread that acquires a
+//! lock in `wgpu-core` will write its own log file to that directory.
+//! You can then run the `wgpu` workspace's `analyze-locks` binary
+//! to read those files and summarize the results. The output from
+//! `analyze-locks` has the same form as the lock ranks given in
+//! [`lock/rank.rs`].
+//!
+//! If the `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is not
+//! set, then no instrumentation takes place, and the locks behave
+//! normally.
+//!
+//! To make sure we capture all acquisitions regardless of when the
+//! program exits, each thread writes events directly to its log file
+//! as they occur. A `write` system call is generally just a copy from
+//! userspace into the kernel's buffer, so hopefully this approach
+//! will still have tolerable performance.
+//!
+//! [`lock/rank.rs`]: ../../../src/lock/rank.rs
+
+#![allow(dead_code)]
```

Comment:
Thanks - I think I just got exhausted and didn't look over the details before throwing the PR up. I believe this attribute is entirely unnecessary because, since this module depends on serde derive macros, it's only compiled when we are indeed using it.